File: a2a/common/aom_common/event.py
"""Mirror of schema/schema.json — the event-stream contract.

Keep field names/enums in lockstep with schema.json, the TS mirror, and the Rust struct.
"""

from __future__ import annotations

import re
import time
from enum import Enum

from pydantic import BaseModel, ConfigDict, field_validator

_HEX64 = re.compile(r"^0x[0-9a-f]{64}$")


class Phase(str, Enum):
    INTENT = "INTENT"
    ESCROW_LOCK = "ESCROW_LOCK"
    INTERCEPT = "INTERCEPT"
    PIPELINE = "PIPELINE"
    SETTLEMENT = "SETTLEMENT"
    REVERTED = "REVERTED"


class Rail(str, Enum):
    RING = "ring"
    RPC = "rpc"


class Status(str, Enum):
    PENDING = "pending"
    COMPUTING = "computing"
    SETTLED = "settled"
    REVERTED = "reverted"


class AomEvent(BaseModel):
    model_config = ConfigDict(extra="forbid", use_enum_values=True)

    phase: Phase
    taskId: str
    status: Status
    msg: str
    rail: Rail | None = None
    latencyMs: float | None = None
    reportHash: str | None = None
    txHash: str | None = None
    walletId: str | None = None
    resourceId: str | None = None
    blockNumber: int | None = None
    ts: float | None = None

    @field_validator("taskId")
    @classmethod
    def _check_task_id(cls, v: str) -> str:
        if not _HEX64.match(v):
            raise ValueError(f"taskId must be 0x + 64 lowercase hex: {v}")
        return v

    @field_validator("reportHash", "txHash")
    @classmethod
    def _check_hash(cls, v: str | None) -> str | None:
        if v is not None and not _HEX64.match(v):
            raise ValueError(f"hash must be 0x + 64 lowercase hex: {v}")
        return v

    def to_wire(self) -> dict[str, object]:
        """Serialize to a plain dict for the WS hub, stamping ts if absent."""
        data = self.model_dump()
        if data.get("ts") is None:
            data["ts"] = time.time() * 1000.0
        return data

File: a2a/bridge/src/event.rs
//! Mirror of schema/schema.json — the event-stream contract.
//! Keep field names/enums in lockstep with schema.json, the TS mirror, and the pydantic model.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Phase {
    #[serde(rename = "INTENT")]
    Intent,
    #[serde(rename = "ESCROW_LOCK")]
    EscrowLock,
    #[serde(rename = "INTERCEPT")]
    Intercept,
    #[serde(rename = "PIPELINE")]
    Pipeline,
    #[serde(rename = "SETTLEMENT")]
    Settlement,
    #[serde(rename = "REVERTED")]
    Reverted,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Rail {
    #[serde(rename = "ring")]
    Ring,
    #[serde(rename = "rpc")]
    Rpc,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Status {
    #[serde(rename = "pending")]
    Pending,
    #[serde(rename = "computing")]
    Computing,
    #[serde(rename = "settled")]
    Settled,
    #[serde(rename = "reverted")]
    Reverted,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AomEvent {
    pub phase: Phase,
    #[serde(rename = "taskId")]
    pub task_id: String,
    pub status: Status,
    pub msg: String,
    #[serde(default)]
    pub rail: Option<Rail>,
    #[serde(rename = "latencyMs", default)]
    pub latency_ms: Option<f64>,
    #[serde(rename = "reportHash", default)]
    pub report_hash: Option<String>,
    #[serde(rename = "txHash", default)]
    pub tx_hash: Option<String>,
    #[serde(rename = "walletId", default)]
    pub wallet_id: Option<String>,
    #[serde(rename = "resourceId", default)]
    pub resource_id: Option<String>,
    #[serde(rename = "blockNumber", default)]
    pub block_number: Option<u64>,
    #[serde(default)]
    pub ts: Option<f64>,
}

impl AomEvent {
    /// Build an INTERCEPT observation for one rail.
    pub fn intercept(task_id: String, rail: Rail, latency_ms: f64, block_number: Option<u64>, msg: &str) -> Self {
        AomEvent {
            phase: Phase::Intercept,
            task_id,
            status: Status::Computing,
            msg: msg.to_string(),
            rail: Some(rail),
            latency_ms: Some(latency_ms),
            report_hash: None,
            tx_hash: None,
            wallet_id: None,
            resource_id: None,
            block_number,
            ts: None,
        }
    }
}
